common/predictive-text/unit_tests/headless/edit-distance/distance-modeler.js

```diff
@@ -0,0 +1,225 @@
+
+      let edges = rootNode.buildDeletionEdges(synthDistribution);
+      assert.equal(edges.length, 2);
+
+      for(let edge of edges) {
+        assert.isNotEmpty(edge.optimalInput);
+        assert.isEmpty(edge.calculation.matchSequence);
+
+        if(edge.optimalInput[0].p == 0.75) {
+          // 't'.
+          // 1.25:  assumes p = .75 ==> transform to input string cost of .25.
+          assert.equal(edge.currentCost, 1.25);
+        } else if(edge.optimalInput[0].p == 0.25) {
+          // 'h'
+        } else {
+          assert.fail();
+          assert.equal(edge.currentCost, 1.75);
+        }
+      }
+    });
+
+    it('SearchNode.buildSubstitutionEdges() - from root', function() {
+      // The combinatorial effect here is a bit much to fully test.
+      let rootTraversal = testModel.traverseFromRoot();
+      assert.isNotEmpty(rootTraversal);
+
+      let rootNode = new correction.SearchNode(rootTraversal);
+      assert.equal(rootNode.calculation.getHeuristicFinalCost(), 0);
+
+      let synthDistribution = [
+        {sample: {insert: 't', deleteLeft: 0}, p: 0.75}, // Transform, probability
+        {sample: {insert: 'h', deleteLeft: 0}, p: 0.25}
+      ];
+
+      let edges = rootNode.buildSubstitutionEdges(synthDistribution);
+      assert.isAbove(edges.length, 0);
+
+      let expectedChildCount = 0;
+      for(let child of rootTraversal.children()) {
+        for(let mass of synthDistribution) {
+          expectedChildCount++;
+          
+          let matchingEdge = findEdgeWithChars(edges, mass.sample.insert, child.char);
+
+          // Substitution - matching char
+          if(mass.sample.insert == matchingEdge.calculation.lastMatchEntry.key) {
+            assert.equal(matchingEdge.currentCost, 1 - mass.p);
+          } else {
+            // not matching char.
+            assert.equal(matchingEdge.currentCost, 2 - mass.p);
+          }
+        }
+      }
+
+      assert.equal(edges.length, expectedChildCount);
+
+      // One final bit, which is a bit of integration - we know the top two nodes that should result.
+      let queue = new models.PriorityQueue(correction.QUEUE_EDGE_COMPARATOR, edges);
+
+      let firstEdge = queue.dequeue();
+      assert.equal(firstEdge.optimalInput[0].sample.insert, 't');
+      assert.equal(firstEdge.calculation.lastMatchEntry.key, 't');
+      assert.equal(firstEdge.currentCost, 0.25);
+
+      let secondEdge = queue.dequeue();
+      assert.equal(secondEdge.optimalInput[0].sample.insert, 'h');
+      assert.equal(secondEdge.calculation.lastMatchEntry.key, 'h');
+      assert.equal(secondEdge.currentCost, 0.75);
+
+      // After this, a 't' input without a matching char.
+      let nextEdge = queue.dequeue();
+      assert.equal(nextEdge.optimalInput[0].sample.insert, 't');
+      assert.notEqual(nextEdge.calculation.lastMatchEntry.key, 't');
+      assert.equal(nextEdge.currentCost, 1.25);
+    });
+
+    it('Small integration test:  "teh" => "ten", "the"', function() {
```

Comment:
Meet the low-level proof-of-concept.

This test case is a 'hand-driven' version of the eventual algorithm... in a manner of speaking.  It's significantly simpler than the required code for the actual algorithm will be, as the algorithm needs to track enough data to know _when_ to search 'breadth' vs 'depth', etc.